File: docusearch/storage.py
"""
Main document storage class.
"""

from __future__ import annotations


import json
import math
import re
import uuid
from collections import Counter
from pathlib import Path
from collections.abc import MutableMapping, Optional, Sequence, Tuple

from .index import ForwardIndex
from .trie import Trie


def generate_doc_id() -> str:
    """Generate a unique document ID"""
    return f"doc_{uuid.uuid4()}"


class DocumentStorage:
    """Searchable document storage"""

    def __init__(self):
        self.trie = Trie()
        self._forward_index = ForwardIndex()
        self._doc_id_to_document: MutableMapping[str, str] = {}
        self._total_documents = 0

    def add_document_from_path(self, file_path: str) -> Sequence[str]:
        """Add a document from a file path or all files in a directory

        Args:
            file_path: Path to a file or directory

        Returns:
            List of document IDs that were added
        """
        path = Path(file_path)
        if not path.exists():
            raise FileNotFoundError(f"Path not found: {file_path}")

        if path.is_file():
            return [self._add_single_file(path)]
        elif path.is_dir():
            return self._add_directory(path)
        else:
            raise ValueError(f"Path is neither a file nor directory: {file_path}")

    def _add_single_file(self, file_path: Path) -> str:
        """Add a single file to the storage"""
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()
        except UnicodeDecodeError:
            with open(file_path, "r", encoding="latin-1") as f:
                content = f.read()

        return self.add_document(content, str(file_path))

    def _add_directory(self, dir_path: Path) -> Sequence[str]:
        """Add all files in a directory to the storage"""
        added_docs = []

        text_extensions = {
            ".txt",
            ".md",
            ".py",
            ".js",
            ".html",
            ".css",
            ".json",
            ".xml",
            ".csv",
            ".tsv",
            ".log",
            ".rst",
            ".tex",
            ".adoc",
            ".org",
        }

        for file_path in dir_path.rglob("*"):
            if file_path.is_file() and file_path.suffix.lower() in text_extensions:
                try:
                    doc_id = self._add_single_file(file_path)
                    added_docs.append(doc_id)
                except Exception as e:
                    print(f"Warning: Could not add {file_path}: {e}")

        return added_docs

    def add_document(self, content: str, doc_id: Optional[str] = None) -> str:
        """Add a document with given content"""
        if doc_id is not None and doc_id in self._doc_id_to_document:
            raise ValueError(f"Document with ID {doc_id} already exists")

        doc_id = generate_doc_id() if doc_id is None else doc_id

        word_counts = Counter(self._tokenize(content))

        self._doc_id_to_document[doc_id] = content

        self._forward_index.add_document(doc_id, word_counts)

        for word, count in word_counts.items():
            if not self.trie.search(word):
                self.trie.insert(word)
            self.trie.add_document_to_word(word, doc_id, count)

        self._total_documents += 1
        return doc_id

    def remove_document(self, doc_id: str) -> bool:
        """Remove a document from storage"""
        if doc_id not in self._doc_id_to_document:
            return False

        word_counts = self._forward_index.get_document_words(doc_id)

        self._forward_index.remove_document(doc_id)

        for word in word_counts:
            self.trie.remove_document_from_word(word, doc_id)

        del self._doc_id_to_document[doc_id]

        self.trie.cleanup_empty_words()

        self._total_documents = max(0, self._total_documents - 1)
        return True

    def search(self, query: str, top_k: int = 5) -> Sequence[Tuple[str, float, str]]:
        """
        Search for documents using TF-IDF scoring

        Returns:
            List of tuples (doc_id, score, content_preview)
        """
        query_words = self._tokenize(query.lower())
        if not query_words:
            return []

        doc_scores: MutableMapping[str, float] = {}

        for word in query_words:
            # Get documents containing this word
            docs_with_word = self.trie.get_documents_for_word(word)

            for doc_id in docs_with_word:
                tf_idf = self._calculate_tf_idf(doc_id, word)

                doc_scores[doc_id] = doc_scores.get(doc_id, 0) + tf_idf

        sorted_docs = sorted(doc_scores.items(), key=lambda x: x[1], reverse=True)

        results = []
        for doc_id, score in sorted_docs[:top_k]:
            content = self._doc_id_to_document.get(doc_id, "")
            preview = self._get_content_preview(content, query_words)
            results.append((doc_id, score, preview))

        return results

    def search_by_prefix(
        self, prefix: str, top_k: int = 5
    ) -> Sequence[Tuple[str, float, str]]:
        """
        Search for documents using prefix matching on query terms

        Returns:
            List of tuples (doc_id, score, content_preview)
        """
        if not prefix.strip():
            return []

        docs_with_prefix = self.trie.get_documents_for_prefix(prefix.lower())

        if not docs_with_prefix:
            return []

        doc_scores: MutableMapping[str, float] = {}

        for doc_id, total_count in docs_with_prefix.items():
            doc_length = self._forward_index.get_document_length(doc_id)
            if doc_length > 0:
                doc_scores[doc_id] = total_count / doc_length

        # Sort by score and return top-k results
        sorted_docs = sorted(doc_scores.items(), key=lambda x: x[1], reverse=True)

        results = []
        for doc_id, score in sorted_docs[:top_k]:
            content = self._doc_id_to_document.get(doc_id, "")
            preview = self._get_content_preview(content, [prefix])
            results.append((doc_id, score, preview))

        return results

    def prefix_search(self, prefix: str) -> List[str]:
        """Search for words that start with the given prefix"""
        return self.trie.starts_with(prefix)

    def get_document_info(self, doc_id: str) -> Optional[MutableMapping]:
        """Get information about a specific document"""
        if doc_id not in self._doc_id_to_document:
            return None

        word_counts = self._forward_index.get_document_words(doc_id)
        doc_length = self._forward_index.get_document_length(doc_id)

        return {
            "doc_id": doc_id,
            "content": self._doc_id_to_document[doc_id],
            "word_counts": word_counts,
            "total_words": doc_length,
            "unique_words": len(word_counts),
        }

    def get_stats(self) -> MutableMapping:
        """Get statistics about the document storage"""
        return {
            "total_documents": len(self._doc_id_to_document),
            "total_words": len(self.trie.get_all_words()),
            "total_documents_in_index": self._total_documents,
        }

    def _calculate_tf_idf(self, doc_id: str, word: str) -> float:
        """Calculate TF-IDF score for a word in a document"""
        tf = self._forward_index.get_tf(doc_id, word)
        doc_freq = self.trie.get_document_frequency(word)
        if doc_freq == 0:
            return 0
        idf = math.log2((self._total_documents + 1) / (doc_freq + 1)) + 1

        return tf * idf

    def _tokenize(self, text: str) -> Iterable[str]:
        """Tokenize text into words"""
        return (
            word for word in re.findall(r"\b[a-zA-Z]+\b", text.lower()) if len(word) > 1
        )

    def _get_content_preview(
        self, content: str, query_words: List[str], max_length: int = 200
    ) -> str:
        """Generate a preview of the content highlighting query words"""
        if len(content) <= max_length:
            return content

        content_lower = content.lower()
        first_pos = len(content)

        for word in query_words:
            pos = content_lower.find(word)
            if pos != -1 and pos < first_pos:
                first_pos = pos

        start = max(0, first_pos - 50)
        end = min(len(content), start + max_length)

        preview = content[start:end]

        if start > 0:
            preview = "..." + preview
        if end < len(content):
            preview = preview + "..."

        return preview

    def smart_search(self, query: str, top_k: int = 5) -> List[Tuple[str, float, str]]:
        r"""
        Smart search that automatically chooses between exact and prefix search

        Rules:
        - If query ends with *, use prefix search (removing the *)
        - Otherwise use exact word matching
        - Interpret \* as literal * (escape the wildcard)

        Returns:
            List of tuples (doc_id, score, content_preview)
        """
        if not query.strip():
            return []

        query = query.replace("\\*", "___ESCAPED_ASTERISK___")

        if query.endswith("*"):
            prefix = query[:-1].strip()  # Remove the *
            if prefix:  # Only search if there's a prefix
                return self.search_by_prefix(prefix, top_k)
            return []

        query = query.replace("___ESCAPED_ASTERISK___", "*")

        return self.search(query, top_k)

    def save(self, file_path: Path) -> None:
        with open(file_path, "w") as f:
            json.dump(
                {
                    "documents": self._doc_id_to_document,
                    "doc_counter": self._doc_counter,
                    "total_documents": self._total_documents,
                    "forward_index": {
                        "documents": self._forward_index._doc_id_to_document,
                        "doc_lengths": self._forward_index._doc_id_to_doc_length,
                    },
                },
                f,
                indent=2,
            )

    @classmethod
    def load(cls, file_path: Path) -> "DocumentStorage":
        with open(file_path, "r") as f:
            data = json.load(f)

        storage = cls(
            documents=data["documents"],
            doc_counter=data["doc_counter"],
            total_documents=data["total_documents"],
            forward_index=ForwardIndex(
                documents=data["forward_index"]["documents"],
                doc_lengths=data["forward_index"]["doc_lengths"],
            ),
        )

        for doc_id, word_counts in storage._forward_index._doc_id_to_document.items():
            for word, count in word_counts.items():
                if not storage.trie.search(word):
                    # TODO: Use a bloom filter?
                    storage.trie.insert(word)
                storage.trie.add_document_to_word(word, doc_id, count)

        return storage

File: pkg/storage/storage.go
package storage

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"unicode"

	"docusearch/pkg/index"
	"docusearch/pkg/trie"
	"github.com/google/uuid"
)

// DocumentInfo represents information about a document
type DocumentInfo struct {
	DocID       string            `json:"doc_id"`
	Content     string            `json:"content"`
	WordCounts  map[string]int    `json:"word_counts"`
	TotalWords  int               `json:"total_words"`
	UniqueWords int               `json:"unique_words"`
}

// Stats represents storage statistics
type Stats struct {
	TotalDocuments        int `json:"total_documents"`
	TotalWords           int `json:"total_words"`
	TotalDocumentsInIndex int `json:"total_documents_in_index"`
}

// SearchResult represents a search result
type SearchResult struct {
	DocID   string  `json:"doc_id"`
	Score   float64 `json:"score"`
	Preview string  `json:"preview"`
}

// StorageData represents the serializable data for persistence
type StorageData struct {
	Documents      map[string]string            `json:"documents"`
	TotalDocuments int                          `json:"total_documents"`
	ForwardIndex   ForwardIndexData             `json:"forward_index"`
}

// ForwardIndexData represents the serializable forward index data
type ForwardIndexData struct {
	Documents  map[string]map[string]int `json:"documents"`
	DocLengths map[string]int            `json:"doc_lengths"`
}

// DocumentStorage provides in-memory document storage with TF-IDF search capabilities
type DocumentStorage struct {
	trie           *trie.Trie
	forwardIndex   *index.ForwardIndex
	docIDToDocument map[string]string
	totalDocuments int
}

// generateDocID generates a unique document ID
func generateDocID() string {
	return fmt.Sprintf("doc_%s", uuid.New().String())
}

// New creates a new DocumentStorage instance
func New() *DocumentStorage {
	return &DocumentStorage{
		trie:           trie.New(),
		forwardIndex:   index.NewForwardIndex(),
		docIDToDocument: make(map[string]string),
		totalDocuments: 0,
	}
}

// NewWithData creates a DocumentStorage instance with existing data
func NewWithData(documents map[string]string, totalDocuments int, forwardIndexData *index.ForwardIndex) *DocumentStorage {
	storage := &DocumentStorage{
		trie:           trie.New(),
		forwardIndex:   forwardIndexData,
		docIDToDocument: documents,
		totalDocuments: totalDocuments,
	}

	// Rebuild trie from forward index
	for docID, wordCounts := range forwardIndexData.GetDocuments() {
		for word, count := range wordCounts {
			if !storage.trie.Search(word) {
				storage.trie.Insert(word)
			}
			storage.trie.AddDocumentToWord(word, docID, count)
		}
	}

	return storage
}

// AddDocumentFromPath adds a document from a file path or all files in a directory
func (ds *DocumentStorage) AddDocumentFromPath(filePath string) ([]string, error) {
	info, err := os.Stat(filePath)
	if err != nil {
		return nil, fmt.Errorf("path not found: %s", filePath)
	}

	if info.IsDir() {
		return ds.addDirectory(filePath)
	} else {
		docID, err := ds.addSingleFile(filePath)
		if err != nil {
			return nil, err
		}
		return []string{docID}, nil
	}
}

// addSingleFile adds a single file to the storage
func (ds *DocumentStorage) addSingleFile(filePath string) (string, error) {
	content, err := ds.readFileContent(filePath)
	if err != nil {
		return "", err
	}

	return ds.AddDocument(content, filePath), nil
}

// readFileContent reads file content with fallback encoding
func (ds *DocumentStorage) readFileContent(filePath string) (string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", err
	}

	// Try UTF-8 first, then fallback to Latin-1 equivalent
	content := string(data)
	if !isValidUTF8(content) {
		// For Go, we'll just use the bytes as-is since Go handles encoding more gracefully
		content = string(data)
	}

	return content, nil
}

// isValidUTF8 checks if the string is valid UTF-8
func isValidUTF8(s string) bool {
	for _, r := range s {
		if r == unicode.ReplacementChar {
			return false
		}
	}
	return true
}

// addDirectory adds all files in a directory to the storage
func (ds *DocumentStorage) addDirectory(dirPath string) ([]string, error) {
	var addedDocs []string

	textExtensions := map[string]bool{
		".txt": true, ".md": true, ".py": true, ".js": true, ".html": true,
		".css": true, ".json": true, ".xml": true, ".csv": true, ".tsv": true,
		".log": true, ".rst": true, ".tex": true, ".adoc": true, ".org": true,
		".go": true, // Adding Go files since this is a Go project
	}

	err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if info.IsDir() {
			return nil
		}

		ext := strings.ToLower(filepath.Ext(path))
		if textExtensions[ext] {
			docID, err := ds.addSingleFile(path)
			if err != nil {
				fmt.Printf("Warning: Could not add %s: %v\n", path, err)
				return nil // Continue processing other files
			}
			addedDocs = append(addedDocs, docID)
		}

		return nil
	})

	return addedDocs, err
}

// AddDocument adds a document with given content
func (ds *DocumentStorage) AddDocument(content, docID string) string {
	if docID == "" {
		docID = generateDocID()
	} else if _, exists := ds.docIDToDocument[docID]; exists {
		panic(fmt.Sprintf("Document with ID %s already exists", docID))
	}

	wordCounts := ds.tokenize(content)
	ds.docIDToDocument[docID] = content
	ds.forwardIndex.AddDocument(docID, wordCounts)

	for word, count := range wordCounts {
		if !ds.trie.Search(word) {
			ds.trie.Insert(word)
		}
		ds.trie.AddDocumentToWord(word, docID, count)
	}

	ds.totalDocuments++
	return docID
}

// RemoveDocument removes a document from storage
func (ds *DocumentStorage) RemoveDocument(docID string) bool {
	if _, exists := ds.docIDToDocument[docID]; !exists {
		return false
	}

	wordCounts := ds.forwardIndex.GetDocumentWords(docID)
	ds.forwardIndex.RemoveDocument(docID)

	for word := range wordCounts {
		ds.trie.RemoveDocumentFromWord(word, docID)
	}

	delete(ds.docIDToDocument, docID)
	ds.trie.CleanupEmptyWords()

	if ds.totalDocuments > 0 {
		ds.totalDocuments--
	}

	return true
}

// Search performs TF-IDF search for documents
func (ds *DocumentStorage) Search(query string, topK int) []SearchResult {
	queryWords := ds.tokenizeQuery(strings.ToLower(query))
	if len(queryWords) == 0 {
		return []SearchResult{}
	}

	docScores := make(map[string]float64)

	for _, word := range queryWords {
		docsWithWord := ds.trie.GetDocumentsForWord(word)

		for docID := range docsWithWord {
			tfIdf := ds.calculateTFIDF(docID, word)
			docScores[docID] += tfIdf
		}
	}

	// Sort documents by score
	type docScore struct {
		docID string
		score float64
	}

	var sortedDocs []docScore
	for docID, score := range docScores {
		sortedDocs = append(sortedDocs, docScore{docID, score})
	}

	sort.Slice(sortedDocs, func(i, j int) bool {
		return sortedDocs[i].score > sortedDocs[j].score
	})

	// Build results
	var results []SearchResult
	limit := topK
	if limit > len(sortedDocs) {
		limit = len(sortedDocs)
	}

	for i := 0; i < limit; i++ {
		docID := sortedDocs[i].docID
		score := sortedDocs[i].score
		content := ds.docIDToDocument[docID]
		preview := ds.getContentPreview(content, queryWords, 200)
		
		results = append(results, SearchResult{
			DocID:   docID,
			Score:   score,
			Preview: preview,
		})
	}

	return results
}

// SearchByPrefix searches for documents using prefix matching on query terms
func (ds *DocumentStorage) SearchByPrefix(prefix string, topK int) []SearchResult {
	if strings.TrimSpace(prefix) == "" {
		return []SearchResult{}
	}

	docsWithPrefix := ds.trie.GetDocumentsForPrefix(strings.ToLower(prefix))
	if len(docsWithPrefix) == 0 {
		return []SearchResult{}
	}

	docScores := make(map[string]float64)

	for docID, totalCount := range docsWithPrefix {
		docLength := ds.forwardIndex.GetDocumentLength(docID)
		if docLength > 0 {
			docScores[docID] = float64(totalCount) / float64(docLength)
		}
	}

	// Sort by score
	type docScore struct {
		docID string
		score float64
	}

	var sortedDocs []docScore
	for docID, score := range docScores {
		sortedDocs = append(sortedDocs, docScore{docID, score})
	}

	sort.Slice(sortedDocs, func(i, j int) bool {
		return sortedDocs[i].score > sortedDocs[j].score
	})

	// Build results
	var results []SearchResult
	limit := topK
	if limit > len(sortedDocs) {
		limit = len(sortedDocs)
	}

	for i := 0; i < limit; i++ {
		docID := sortedDocs[i].docID
		score := sortedDocs[i].score
		content := ds.docIDToDocument[docID]
		preview := ds.getContentPreview(content, []string{prefix}, 200)
		
		results = append(results, SearchResult{
			DocID:   docID,
			Score:   score,
			Preview: preview,
		})
	}

	return results
}

// PrefixSearch searches for words that start with the given prefix
func (ds *DocumentStorage) PrefixSearch(prefix string) []string {
	return ds.trie.StartsWith(prefix)
}

// GetDocumentInfo gets information about a specific document
func (ds *DocumentStorage) GetDocumentInfo(docID string) *DocumentInfo {
	content, exists := ds.docIDToDocument[docID]
	if !exists {
		return nil
	}

	wordCounts := ds.forwardIndex.GetDocumentWords(docID)
	docLength := ds.forwardIndex.GetDocumentLength(docID)

	return &DocumentInfo{
		DocID:       docID,
		Content:     content,
		WordCounts:  wordCounts,
		TotalWords:  docLength,
		UniqueWords: len(wordCounts),
	}
}

// GetStats gets statistics about the document storage
func (ds *DocumentStorage) GetStats() Stats {
	return Stats{
		TotalDocuments:        len(ds.docIDToDocument),
		TotalWords:           len(ds.trie.GetAllWords()),
		TotalDocumentsInIndex: ds.totalDocuments,
	}
}

// SmartSearch automatically chooses between exact and prefix search
func (ds *DocumentStorage) SmartSearch(query string, topK int) []SearchResult {
	if strings.TrimSpace(query) == "" {
		return []SearchResult{}
	}

	// Handle escaped asterisks
	query = strings.ReplaceAll(query, "\\*", "___ESCAPED_ASTERISK___")

	if strings.HasSuffix(query, "*") {
		prefix := strings.TrimSpace(strings.TrimSuffix(query, "*"))
		if prefix != "" {
			return ds.SearchByPrefix(prefix, topK)
		}
		return []SearchResult{}
	}

	// Restore escaped asterisks
	query = strings.ReplaceAll(query, "___ESCAPED_ASTERISK___", "*")

	return ds.Search(query, topK)
}

// calculateTFIDF calculates TF-IDF score for a word in a document
func (ds *DocumentStorage) calculateTFIDF(docID, word string) float64 {
	tf := ds.forwardIndex.GetTF(docID, word)
	docFreq := ds.trie.GetDocumentFrequency(word)
	if docFreq == 0 {
		return 0
	}
	idf := math.Log2(float64(ds.totalDocuments+1)/float64(docFreq+1)) + 1

	return tf * idf
}

// tokenize tokenizes text into words
func (ds *DocumentStorage) tokenize(text string) map[string]int {
	wordRegex := regexp.MustCompile(`\b[a-zA-Z]+\b`)
	words := wordRegex.FindAllString(strings.ToLower(text), -1)
	
	wordCounts := make(map[string]int)
	for _, word := range words {
		if len(word) > 1 { // Only include words longer than 1 character
			wordCounts[word]++
		}
	}
	
	return wordCounts
}

// tokenizeQuery tokenizes query text into words
func (ds *DocumentStorage) tokenizeQuery(text string) []string {
	wordRegex := regexp.MustCompile(`\b[a-zA-Z]+\b`)
	words := wordRegex.FindAllString(text, -1)
	
	var result []string
	for _, word := range words {
		if len(word) > 1 {
			result = append(result, word)
		}
	}
	
	return result
}

// getContentPreview generates a preview of the content highlighting query words
func (ds *DocumentStorage) getContentPreview(content string, queryWords []string, maxLength int) string {
	if len(content) <= maxLength {
		return content
	}

	contentLower := strings.ToLower(content)
	firstPos := len(content)

	for _, word := range queryWords {
		pos := strings.Index(contentLower, word)
		if pos != -1 && pos < firstPos {
			firstPos = pos
		}
	}

	start := 0
	if firstPos < len(content) {
		start = firstPos - 50
		if start < 0 {
			start = 0
		}
	}

	end := start + maxLength
	if end > len(content) {
		end = len(content)
	}

	preview := content[start:end]

	if start > 0 {
		preview = "..." + preview
	}
	if end < len(content) {
		preview = preview + "..."
	}

	return preview
}

// Save saves the storage to a JSON file
func (ds *DocumentStorage) Save(filePath string) error {
	data := StorageData{
		Documents:      ds.docIDToDocument,
		TotalDocuments: ds.totalDocuments,
		ForwardIndex: ForwardIndexData{
			Documents:  ds.forwardIndex.GetDocuments(),
			DocLengths: ds.forwardIndex.GetDocLengths(),
		},
	}

	file, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer file.Close()

	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	return encoder.Encode(data)
}

// Load loads storage from a JSON file
func Load(filePath string) (*DocumentStorage, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	data, err := io.ReadAll(file)
	if err != nil {
		return nil, err
	}

	var storageData StorageData
	if err := json.Unmarshal(data, &storageData); err != nil {
		return nil, err
	}

	forwardIndex := index.NewForwardIndexWithData(
		storageData.ForwardIndex.Documents,
		storageData.ForwardIndex.DocLengths,
	)

	storage := NewWithData(
		storageData.Documents,
		storageData.TotalDocuments,
		forwardIndex,
	)

	return storage, nil
} 